pysal/contrib/viz/color.py

```diff
@@ -0,0 +1,178 @@
+"""
+color handling for mapping and geovisualization
+"""
+
+import matplotlib.pyplot as plt
+import matplotlib.patches as patches
+import matplotlib.colors as mpc
+
+
+try:
+    import brewer2mpl
+    from brewer2mpl import qualitative as b2m_qualitative
+    from brewer2mpl import sequential as b2m_sequential
+    from brewer2mpl import diverging as b2m_diverging
+except:
+    print('brewer2mpl  not installed. Functionality '
+          'related to it will not work')
+try:
+    import palettable as pltt
+    from palettable.colorbrewer import qualitative, sequential, diverging
+except:
+    print('palettable  not installed. Functionality '
+          'related to it will not work')
+
+def get_color_map(palette=None, name='BuGn', cmtype='sequential', k=5,
+                  color_encoding='hex'):
+    """
+    Get a brewer colormap from `palettable`
+
+    Arguments
+    ---------
+
+    palette         : `palettable` palette
+                      Palettable object for a given palette
+    name            : string
+                      colormap name
+
+    cmtype          : string
+                      colormap scale type  [sequential, diverging, qualitative]
+
+    k               : int
+                      number of classes
+
+    color_encoding  : string
+                      encoding of colors [hexc, rgb, mpl, mpl_colormap]
+                      * hex: list of hex strings
+                      * rgb: list of RGB 0-255 triplets
+                      * mpl: list of RGB 0-1 triplets as used by matplotlib
+                      * mpl_colormap: matplotlib color map
+
+    Returns
+    -------
+
+    colors:  color map in the specified color_encoding
+
+    """
+    encs = {'hex': 'hex_colors',
+            'rgb': 'colors',
+            'mpl': 'mpl_colors',
+            'mpl_colormap': 'mpl_colormap'}
+    if not palette:
+        cmtype = pltt2type[name.lower()]
+        if name[-2:] == '_r':
+            palette = pltt.colorbrewer.get_map(name[:-2], cmtype,
+                    k, reverse=True)
+        else:
+            palette = pltt.colorbrewer.get_map(name, cmtype, k)
+    colors = getattr(palette, encs[color_encoding.lower()])
+    return colors
+
+def _build_pltt2type():
+    types = ['sequential', 'diverging', 'qualitative']
+    pltt2type = {}
+    for t in types:
+        pals = list(set([
+            p.split('_')[0] for p in dir(getattr(pltt.colorbrewer, t)) 
+                            if p[0]!='_'
+                            ]))
+        pals = pals + [p+'_r' for p in pals]
+        for p in pals:
+            pltt2type[p.lower()] = t
+    return pltt2type
+
+pltt2type = _build_pltt2type()
```

Comment:
This line causes the module to fail in import when you don't have `palettable`. 

Even though `pltt` is conditionally imported at top, this function gets run with no checks on whether `pltt` exists. This means you get a `pltt is not defined` error when this gets called, since the module is not defined. 

I will go through and tag things with the `@requires` decorator, just in case. I think that's the [safest way to do soft dependent functions/classes with no alternative implementation](https://github.com/pysal/pysal/wiki/Using-soft-dependencies#using-requires).